### ISSUE TITLE
batch size and acceleration ratio

### ISSUE THREAD

**Issue by Emily-Ward** [mobiusml/gemlite#13]
Why does the acceleration benefit of the quantized operator over the direct fp16 torch.matmul decrease as the batch size increases?
I've seen in many places that the overhead of dequantization results in no acceleration benefit. Is that true?
If dequantization doesn't bring acceleration, why is there an acceleration benefit when the batch size is small? Is this acceleration benefit related to batch size?

**Comment by mobicham** [comment]
You get faster decoding speed with quantized weights when you are memory-bound, that's typically the case for batch-sizes between 1 and 64. For larger batch-sizes - prefill phase, the quantized matmul will actually run slower because of the dequantization overhead - except in some cases (in GemLite we can use fp16 accumulation to make it faster than torch.matmul p16).
In production though, a typically batch size would be something like 8-32-ish, so you still benefit from the speed-up + you save a lot of VRAM and cost when the quantized weights.

**Comment by Emily-Ward** [comment]
If it's in the self-decode phase, will `gemm_A16fWnO16f_int32packing_forward` in GemLite be faster than `torch.matmul fp16`? Or in other words, does the acceleration benefit in the self-decode phase decrease as the batch size increases? For example, for operators like `gate_proj`, `down_proj`, and `up_proj`, if the batch size is 512, what would the situation look like when considering only the self-decode phase?

**Comment by Emily-Ward** [comment]
![image](https://github.com/user-attachments/assets/cd44b74a-46d6-4401-b981-60f57cdfe45e)
![image](https://github.com/user-attachments/assets/ad15a222-d243-44ed-ab09-37c9cb45086b)
Okay, I roughly understand now. In the self-decode phase, increasing the batch size might cause the MLP module to become compute-bound, so the overhead of dequantization limits the acceleration benefit.

**Comment by mobicham** [comment]
There are 3 matmul algorithms in Gemlite: `gemm_A16fWnO16f_int32packing_forward` is used for batch-size>64 - typically the prefill phase. For lower batch-sizes - typically decoding phase, we use the gemv revsplitk for batch-size=1 and the split-K implementation for batch-sizes between 2-64. 
GemLite's gemm will only be faster than torch.matmul fp16 on consumer gpus (3090, 4090) with fp16 accumulation. However, the gemm will mainly be used for the prefill phase. For decoding where we actually need the speed-up, it will either use the gemv revsplitk or the split-k implementation depending on the batch-size.
batch-size 512 is not a realistic batch-size for inference, typical batch size is around 1-32.

**Comment by Emily-Ward** [comment]
Sorry to bother you, I'm trying to optimize the gemm_A16fWnO16f_int32packing_kernel, but I'm unsure about the shape of scales. I noticed a comment saying 'scales and zeros are of shape (group_size, N): float16 or bfloat16', but I'm not sure why it has this shape.

**Comment by Emily-Ward** [comment]
Sorry to bother you, I'm trying to optimize the gemm_A16fWnO16f_int32packing_kernel, but I'm unsure about the shape of scales. I noticed a comment saying 'scales and zeros are of shape (group_size, N): float16 or bfloat16', but I'm not sure why it has this shape.

**Comment by Emily-Ward** [comment]
Now I see, thank you for your help. However, when I run inference, I encountered the following error, and I'm not quite sure what this error means?

![image](https://github.com/user-attachments/assets/db590832-f804-49b0-9994-c209888de887)

**Comment by mobicham** [comment]
Can you share a code snippet ? Looks like your tensor is on cpu. 

It's not recommended to run the kernels in isolation, there's some logic in `core.py` that ensures the kernels get the right inputs and the right configs.

**Comment by Emily-Ward** [comment]
Ok, here:
import torch
import time
from hqq.core.quantize import *
from hqq.models.hf.base import AutoHQQHFModel
from transformers import AutoModelForCausalLM, AutoTokenizer
from hqq.utils.patching import prepare_for_inference
from hqq.utils.generation_hf import patch_model_for_compiled_runtime

#load model
device = 'cuda:0'
cache_dir = '.'
model_id = '/workspace/llama/llama3/Meta-Llama-3-8B'
compute_dtype = torch.float16
backend = 'gemlite'
tokenizer = AutoTokenizer.from_pretrained(model_id, cache_dir=cache_dir)
model = AutoModelForCausalLM.from_pretrained(model_id, cache_dir=cache_dir, torch_dtype=compute_dtype, attn_implementation="sdpa")
if tokenizer.pad_token_id is None:
    tokenizer.pad_token_id = tokenizer.eos_token_id

#quant model
q8_config = BaseQuantizeConfig(nbits=8, group_size=128)
q4_config = BaseQuantizeConfig(nbits=4, group_size=64)
q2_config = BaseQuantizeConfig(nbits=2, group_size=128)
q1_config = BaseQuantizeConfig(nbits=1, group_size=128)

quant_config = {
    'mlp.down_proj': q4_config,
}
#quant_config = {}
AutoHQQHFModel.quantize_model(model, quant_config = quant_config, compute_dtype=compute_dtype, device=device)


#Optimize
from hqq.utils.patching import prepare_for_inference
prepare_for_inference(model, backend=backend, verbose=True)


WARMUP_PROMPTS = [
    "Write an essay about large language models.",
    #"Tell me a funny joke!",
    #"Hello, my name is Kang, I like England for five reasons. First,",
    #"Who is Elon Musk?",
]
for i in range(2):
    inputs_exec = tokenizer(WARMUP_PROMPTS*1,return_tensors='pt',padding='max_length',max_length=128,truncation=True,padding_side='left').to(model.device)
    input_token_length = inputs_exec['input_ids'].shape[1]
    bs = inputs_exec['input_ids'].shape[0]
    print('batch_size: ',bs)
    print('seq len: ',input_token_length)
    # Generate
    torch.cuda.synchronize()
    exec_start = time.time()
    output = model.generate(**inputs_exec, max_new_tokens=1, cache_implementation="static", pad_token_id=tokenizer.pad_token_id,do_sample=False,temperature=None,top_p=None) 
    torch.cuda.synchronize()
    exec_end = time.time()
    output_token_length = output.shape[1]
    print('exec : ', exec_end - exec_start,'\t new tokens: ',output_token_length-input_token_length,'\t tokens/s ',(output_token_length-input_token_length)/(exec_end-exec_start)*bs)

**Comment by mobicham** [comment]
This one works for sure as of yesterday
```Python
import torch
device        = 'cuda:0'
backend       = "gemlite" #'torchao_int4' #"torchao_int4" (4-bit only) or "bitblas" (4-bit + 2-bit) or "gemlite" (8-bit, 4-bit, 2-bit, 1-bit)
compute_dtype = torch.bfloat16 if backend=="torchao_int4" else torch.float16
cache_dir     = '.' 
model_id      = 'meta-llama/Meta-Llama-3-8B-Instruct'

########################################################################
#Load model
from transformers import AutoModelForCausalLM, AutoTokenizer
from hqq.models.hf.base import AutoHQQHFModel
from hqq.core.quantize import *

#Load
tokenizer = AutoTokenizer.from_pretrained(model_id, cache_dir=cache_dir)
model     = AutoModelForCausalLM.from_pretrained(model_id, cache_dir=cache_dir, torch_dtype=compute_dtype, attn_implementation="sdpa")

#Quantize
quant_config = BaseQuantizeConfig(nbits=4, group_size=64, axis=1)
AutoHQQHFModel.quantize_model(model, quant_config=quant_config, compute_dtype=compute_dtype, device=device)
HQQLinear.set_backend(HQQBackend.PYTORCH)

#Optimize
from hqq.utils.patching import prepare_for_inference
prepare_for_inference(model, backend=backend, verbose=True)

#Inference
########################################################################
#Using a custom generator
from hqq.utils.generation_hf import HFGenerator
gen = HFGenerator(model, tokenizer, max_new_tokens=1000, do_sample=True, compile="partial").warmup() 

out = gen.generate("Write an essay about large language models.", print_tokens=True)
```

The warm-up the first time will take a long time, but can cache the config after the warm-up. <a href="https://github.com/mobiusml/hqq/blob/master/examples/backends/hqq_lib_demo.py#L32-L35">here's an example</a>.

Make sure you use the master branches:
```
pip install --upgrade git+https://github.com/mobiusml/hqq.git ;
pip install --upgrade git+https://github.com/mobiusml/gemlite.git;
```

**Comment by Emily-Ward** [comment]
The issue is resolved. I accidentally deleted part of the code. Thank you for your help!
